### PR TITLE
[Spells] Major update to cast restriction code and new spell field 'caster_requirement_id' (field220) implemented

### PR DESCRIPTION
**Pull request by KayenEQ**
Updated my code that handles field CastRestriction in spells_new table and the Limit effect on SPA 0 and SPA 79.
This effect restricts if a spell can be applied to a target based on specific conditions.
Many new restriction id's were implemented, many old id's were corrected to reflect better descriptions.

Implemented spells_new table 'field220' as 'caster_requirement_id'. This field is used to apply requirements on the caster before a spell can be cast. Any CastRestriction ID's can be used for this field to give caster same restriction.

Restriction ID corresponds to the type 39 value in dstr_us on live clients (2021). See enum SpellRestriction for full list in spdat.h
Modern client will give a message corresponding the type 39 field in the dstr_us for many of these effects upon failure.
If from CasterRestriction spell field. "Your target does not meet the spell requirements. <insert below corresponding string>." Msg in combat window, color red.

### CODE REVIEW

File: common/spdat.h [hunk 1]
```diff
@@ -189,6 +200,306 @@ enum FocusLimitIncludes {
...
+	IS_PLANT                                                                  = 105, // This spell will only work on plants. 
+	IS_GIANT                                                                  = 106, // This spell will only work on animals. | Live used to have this on spells restricted to Giants, but those spells were removed... We still have them
+	IS_NOT_ANIMAL_OR_HUMANOID                                                 = 108, // This spell will only work on targets that are neither animals or humanoid. 
+	IS_BIXIE                                                                  = 109, // This spell will only work on bixies. 
+	IS_HARPY                                                                  = 110, // This spell will only work on harpies. 
+	IS_GNOLL                                                                  = 111, // This spell will only work on gnolls. 
+	IS_SPORALI                                                                = 112, // This spell will only work on fungusoids. 
+	IS_KOBALD                                                                 = 113, // This spell will only work on kobolds.
```
Comment (Kinglykrab) [inline common/spdat.h:224]: Typo.

File: zone/spell_effects.cpp [hunk 2]
```diff
@@ -7109,331 +7111,948 @@ bool Mob::ImprovedTaunt(){
...
+		case IS_SPORALI:
+			if ((GetRace() == RT_SPORALI) ||(GetRace() == RT_FUNGUSMAN))
+				return true;
+			break;
 
-			case 115:
-				if (GetRace() == 522)
-					return true;
-				break;
+		case IS_KOBALD:
```
Comment (Kinglykrab) [inline zone/spell_effects.cpp:7210]: Typo.

**Review by Kinglykrab — APPROVED** [review]
Looks good, noticed a typo.

File: common/spdat.h (hunk 1 above)
Comment (KayenEQ) [reply common/spdat.h]: fixed

File: zone/spell_effects.cpp (hunk 2 above)
Comment (KayenEQ) [reply zone/spell_effects.cpp]: fixed

File: zone/spell_effects.cpp [hunk 3]
```diff
@@ -7109,331 +7111,948 @@ bool Mob::ImprovedTaunt(){
...
+			if (count > 14) {
 				return true;
+			}
+			break;
 		}
-	} // End Heal
 
+		case IS_CLASS_CHAIN_OR_PLATE:
+			if (IsClient() &&
+				((GetClass() == WARRIOR) || (GetClass() == BARD)  || (GetClass() == SHADOWKNIGHT)  || (GetClass() == PALADIN)  || (GetClass() == CLERIC)
```
Comment (Akkadius) [inline zone/spell_effects.cpp:7484]: This could have been summarized with helpers `IsChain(GetClass())` or `IsPlate(GetClass())`

**Review by Akkadius — APPROVED** [review]
Beautiful work man!

**Pull request merged**